examples/preprocessing/plot_target_encoder.py

```diff
@@ -0,0 +1,120 @@
+
+wine_reviews = fetch_openml(data_id=42074, as_frame=True, parser="liac-arff")
+
+df = wine_reviews.frame
+df.head()
+
+# %%
+# For this example, we use the following subset of numerical and categorical
+# features in the data. The target are continuous values from 80 to 100:
+numerical_features = ["price"]
+categorical_features = [
+    "country",
+    "province",
+    "region_1",
+    "region_2",
+    "variety",
+    "winery",
+]
+target_name = "points"
+
+X = df[numerical_features + categorical_features]
+y = df[target_name]
+
+y.hist()
+
+# %%
+# Then, we split the dataset into a training and test set.
+from sklearn.model_selection import train_test_split
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
+
+print(f"Samples in training set: {len(X_train)}\nSamples in test set: {len(X_test)}")
+
+# %%
+# Training and Evaluating Pipelines with Different Encoders
+# =========================================================
+# In this section, we will evaluate pipelines with
+# :class:`~sklearn.ensemble.HistGradientBoostingRegressor` with different encoding
+# strategies. First, we list out the encoders we will be using to preprocess
+# the categorical features:
+from sklearn.compose import ColumnTransformer
+from sklearn.preprocessing import OrdinalEncoder
+from sklearn.preprocessing import OneHotEncoder
+from sklearn.preprocessing import TargetEncoder
+
+categorical_preprocessors = [
+    ("drop", "drop"),
+    ("ordinal", OrdinalEncoder(handle_unknown="use_encoded_value", unknown_value=-1)),
+    (
+        "one_hot_encoder",
+        OneHotEncoder(handle_unknown="ignore", max_categories=50, sparse_output=False),
+    ),
+    ("target", TargetEncoder(target_type="continuous")),
+]
+
+# %%
+# Next, we create and fit the pipelines to the training data and evaluate them on the
+# test set:
+from sklearn.pipeline import make_pipeline
+from sklearn.ensemble import HistGradientBoostingRegressor
+from sklearn.metrics import mean_squared_error
+
+results = []
+for name, categorical_preprocessor in categorical_preprocessors:
+    preprocessor = ColumnTransformer(
+        [
+            ("nummerical", "passthrough", numerical_features),
+            ("categorical", categorical_preprocessor, categorical_features),
+        ]
+    )
+    pipe = make_pipeline(preprocessor, HistGradientBoostingRegressor(random_state=0))
+    pipe.fit(X_train, y_train)
+
+    y_pred = pipe.predict(X_test)
+    rmse = mean_squared_error(y_test, y_pred, squared=False)
+    results.append({"preprocessor": name, "root mean squared error": rmse})
+
+# %%
+# Finally, we display the results for all the encoders. When evaluating the
+# predictive performance on the test set, dropping the categories perform the
+# worst and the target encoder performs the best. The target encoding provides
+# more information about the target, which the gradient boosted model at the end
+# of the pipeline can take advantage of. In this example, we evaluate with a single
```

Comment:
I think the main explanation is that:

- dropping the categorical features makes the pipeline less expressive and underfitting as a result;
- the one-hot encoding scheme makes the pipeline overfitting as the larger number of features explodes with rare category occurrences that are correlated with the target by chance (on the training set only);
- the ordinal encoding imposes an arbitrary order to the features which are then treated as numerical values by the `HistGradientBoostingRegressor` model. Since this model groups numerical features in 256 bins per feature, many unrelated categories can be grouped together and as a result overall pipeline can underfit.
- when using the target encoder, the same binning happens, but since the encoded values are statistically ordered by marginal association with the target variable, the binning use by the `HistGradientBoostingRegressor` model makes sense and leads to good results: the combination of smoothed target encoding and binning works as a good regularizing strategy against overfitting while not limiting the expressiveness of the pipeline too much.

I think we should extend the plot to also display the training RMSE to confirm the above analysis and make sure it is correct (biggest train / test gap for OHE, train and test much closer for the presumably underfitting models, with the train RMSE constraining the performance). Ideally this would all be with error bars with `cross_validate` together with `return_train_score=True`.